### ISSUE TITLE
mix meta and normal keys

### ISSUE THREAD

**Issue by markus2330** [ElektraInitiative/libelektra#834]
One way to easily distinguish between meta and normal keys would be to use different assignments.

For example:

```
[path]
 key = value
 metakey := metavalue
```

Where key is a subkey of `path` but `metakey` is a metakey of `path`.

@tom-wa What do you think?

**Comment by markus2330** [comment]
@sanssecours Would also be interesting to hear your opinion. Do you already have an idea which plugins with which implementation strategies you want to do?

**Comment by markus2330** [comment]
alternatives for meta assignment would be `<-`.

But maybe its better to use one of the characters which are reserved in the key names (and do not require further escaping), e.g.
- `#=`
- `.=`
- `@=`

**Comment by sanssecours** [comment]
> @sanssecours Would also be interesting to hear your opinion.

I think all the assignment operators you mentioned look fine. Maybe we could even drop the `=` in `:=`? Your example would then look like this:

```
[path]
  key = value
  metakey : metavalue
```

According to [Wikipedia](https://en.wikipedia.org/wiki/INI_file#Name.2Fvalue_delimiter)

> some implementations allow a colon (:) as the name/value delimiter (instead of the equals sign)

, but I never saw such a file.

> Do you already have an idea which plugins with which implementation strategies you want to do?

Not yet, sorry. I just finished reading the book “[Parsing Techniques](http://dickgrune.com/Books/PTAPG_2nd_Edition)”. Now I need to extend and clarify my [exposé](https://www.dropbox.com/sh/3ba8k1kqasa2md4/AAB0oxwHcoKfZhEEezTb8nlXa?dl=0). Sorry for taking so long.

**Comment by markus2330** [comment]
Yes, I also like the `:`.

@tom-wa also suggested that we can distinguish meta data with a special character in the beginning of the line, similar to comments.

Good that you deepen your knowledge about parsing! Thanks for updating the Exposé, I will give you feedback soon!

**Comment by tom-wa** [comment]
ok, i just suggested the special character at the beginning of the line because its easy to implement. the parser became quite a monster (it even parses ascii art) and currently the user can define the delimiter. 
having 2 delimiters would mean we need a stricter syntax and probably wont pass all unit tests

**Comment by tom-wa** [comment]
btw, i'm not against having a stricter syntax. 
currently `"k=e = y" = value`, `key = val1=a val2=b val3=c`, `k=e=y = v=al     ue`, etc will all be parsed correct.
having 2 delimiters would make that (almost ?) impossible. 
so it's either: using a special character at the beginning of the line, or forbid both delimiters withing the keyname
also, does that mean i can drop the "meta-mode" ?

**Comment by markus2330** [comment]
The goal is that we have a consistent behaviour across our parsers, so please do not drop before @sanssecours  decided what he wants to do.

Some way to always have meta without special mode sounds appealing, though.

**Comment by tom-wa** [comment]
i've tried looking into that issue again, adding the metakeys to `ini` seems (almost) impossible with the current legal syntax.

i've been planing to partially rewrite the plugin though (to improve performance and do some refactoring), and introduce a limited the / introduce a stricter syntax. imho it doesn't make much sense for the ini plugin to accept and parse everything like it currently does. and it would allow us to introduce other features more easily (i already have some things in mind)

other applications wouldn't be able to parse most of what's currently allowed anyway (e.g. keyname containing the delimiter character, allowing multiline keynames, etc)

@markus2330 would you be ok with such a stricter/limited syntax ?

**Comment by markus2330** [comment]
I was never in favour of having "heuristics" which = really is the = ;) So go for it. But better not for the next release.

But I think this issue is rather independent of the meta data issue, because you can (and probably should) use comments for meta-data so that non-meta-data aware parsers do not get confused.

**Comment by tom-wa** [comment]
>I was never in favour of having "heuristics" which = really is the = ;) So go for it. But better not for the next release.

i think there should be a way to add some exceptions to the unittests, or just use them to approximate the quality of a storage plugin
i think i'll do it for 0.8.19

  >But I think this issue is rather independent of the meta data issue, because you can (and probably should) use comments for meta-data so that non-meta-data aware parsers do not get confused.

true, but lets assume we use `#@ key = value`, the same syntaxes as for the normal key-value pairs should also be possible here, otherwise it would be just strange.

**Comment by markus2330** [comment]
For plugins capable of *also* storing metadata, I think they should say so in `provides = metastorage` and we should avoid storage plugins not capable of storing normal config.
What do you think?

See also #994 for longer discussion about formats.

**Comment by markus2330** [comment]
Implemented by INI, hopefully by others (XML #1280 via attributes) and YML (@sanssecours via tags?), soon too.

**Comment by sanssecours** [comment]
> @sanssecours via tags?

Though I am not sure how to exactly implement this feature yet, specifying metadata via a local tag (e.g. `!meta`) sounds like a good idea.
